### PR TITLE
feat: replace kwargs with invocation_state in agent APIs

### PR DESCRIPTION
**Pull request by JackYPCOnline**
## Description

Add invocation_state as parameter to agent's APIs. 
Still keep **kwargs for now to keep backward compatible, merge **kwargs to invocation_state instead.

## Related Issues

https://github.com/strands-agents/sdk-python/issues/919
## Type of Change

Bug fix
New feature
Breaking change
Documentation update
Other (please describe):

## Testing

How have you tested the change?  Verify that the changes do not break functionality or introduce warnings in consuming repositories: agents-docs, agents-tools, agents-cli

- [x] I ran `hatch run prepare`

## Checklist
- [x ] I have read the CONTRIBUTING document
- [x ] I have added any necessary tests that prove my fix is effective or my feature works
- [ ] I have updated the documentation accordingly
- [ ] I have added an appropriate example to the documentation to outline the feature, or no new docs are needed
- [x ] My changes generate no new warnings
- [x ] Any dependent changes have been merged and published

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/strands/agent/agent.py [hunk 1]
```diff
@@ -389,6 +392,7 @@ def __call__(self, prompt: AgentInput = None, **kwargs: Any) -> AgentResult:
                 - list[ContentBlock]: Multi-modal content blocks
                 - list[Message]: Complete messages with roles
                 - None: Use existing conversation history
+            invocation_state: [New] Additional parameters to pass through the event loop.
```
Comment (pgrayy) [inline src/strands/agent/agent.py]: Nit: I would recommend getting rid of the [New] label since at some point it will no longer be new.

File: src/strands/agent/agent.py [hunk 2]
```diff
@@ -399,15 +403,24 @@ def __call__(self, prompt: AgentInput = None, **kwargs: Any) -> AgentResult:
                 - metrics: Performance metrics from the event loop
                 - state: The final state of the event loop
         """
+        if kwargs:
+            logger.warning("`Kwargs` parameter is deprecated, use the `invocation_state` parameter instead.")
+            self.invocation_state = kwargs
```
Comment (pgrayy) [inline src/strands/agent/agent.py]: Does invocation_state need to be a member variable?

File: src/strands/agent/agent.py [hunk 3]
```diff
@@ -399,15 +403,24 @@ def __call__(self, prompt: AgentInput = None, **kwargs: Any) -> AgentResult:
                 - metrics: Performance metrics from the event loop
                 - state: The final state of the event loop
         """
+        if kwargs:
```
Comment (pgrayy) [inline src/strands/agent/agent.py]: I would recommend handling this check just once in `stream_async`.

File: src/strands/agent/agent.py [hunk 4]
```diff
@@ -399,15 +403,24 @@ def __call__(self, prompt: AgentInput = None, **kwargs: Any) -> AgentResult:
...
+            logger.warning("`Kwargs` parameter is deprecated, use the `invocation_state` parameter instead.")
+            self.invocation_state = kwargs
+            if invocation_state is not None:
+                self.invocation_state["invocation_state"] = invocation_state
```
Comment (pgrayy) [inline src/strands/agent/agent.py]: Do we parse the "invocation_state" key somewhere? How will this be used?

File: src/strands/agent/agent.py (hunk 2 above)
Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: Just reference the [sudo code](https://github.com/strands-agents/sdk-python/issues/919) provides,  but yes, I think local varibales could prevent race condition in concurrent call?

File: src/strands/agent/agent.py (hunk 4 above)
Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: There is two way to handle :  1. dict.update make it plan dict 2. make it nested. This is idea from https://github.com/strands-agents/sdk-python/issues/919

File: src/strands/agent/agent.py (hunk 3 above)
Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: I see. Make sense.

File: src/strands/agent/agent.py [hunk 5]
```diff
@@ -432,7 +446,15 @@ async def invoke_async(self, prompt: AgentInput = None, **kwargs: Any) -> AgentR
...
-        events = self.stream_async(prompt, **kwargs)
+        if kwargs:
+            logger.warning("`Kwargs` parameter is deprecated, use the `invocation_state` parameter instead.")
+            self.invocation_state = kwargs
+            if invocation_state is not None:
+                self.invocation_state["invocation_state"] = invocation_state
+        else:
+            self.invocation_state = invocation_state
+
+        events = self.stream_async(prompt, invocation_state=self.invocation_state)
```
Comment (Unshure) [inline src/strands/agent/agent.py]: Lets put all of this logic in one place so we dont need to keep repeating it

**Review by Unshure — CHANGES_REQUESTED** [review]

File: src/strands/agent/agent.py (hunk 5 above)
Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: follow @pgrayy 's comment, pass `**kwargs`, only handle it in `stream_async`

File: src/strands/agent/agent.py [hunk 6]
```diff
@@ -567,7 +572,21 @@ async def stream_async(
...
-        callback_handler = kwargs.get("callback_handler", self.callback_handler)
+        merged_state = {}
+        if kwargs:
+            logger.warning("`**kwargs` parameter is deprecated, use `invocation_state` instead.")
+            merged_state.update(kwargs)
+            if invocation_state is not None:
+                merged_state["invocation_state"] = invocation_state
+        else:
+            if invocation_state is not None:
+                merged_state["invocation_state"] = invocation_state
```
Comment (mkmeral) [inline src/strands/agent/agent.py]: aren't we double nesting here? 

```
4. Backward compatibility breaks silently
Old code using agent("prompt", user_id="123") will work but 
with a deprecation warning. However, if someone tries to migrate 
by wrapping it: agent("prompt", invocation_state={"user_id": "123"}), 
the data ends up in a different location in the state dict.
```

so when you try to use `agent("prompt", invocation_state={"user_id": "123"})`, you would need to do this in the tool:
```
tool_context.invocation_state.invocation_state.user_id
```

is this the desired devx for now?

Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: yep, see [comment](https://github.com/strands-agents/sdk-python/pull/966#discussion_r2400049494)

**Comment by JackYPCOnline** [comment]
> Can you add unit tests for this

Actually this change is covered by origin test code but I can add a specific one

File: src/strands/agent/agent.py [hunk 7]
```diff
@@ -567,7 +572,21 @@ async def stream_async(
...
+
+        # Get callback handler from merged state or use default
+        invocation_state_dict = merged_state.get("invocation_state") or {}
```
Comment (afarntrog) [inline src/strands/agent/agent.py]: nit: `merged_state.get("invocation_state", {})`

File: src/strands/agent/agent.py [hunk 8]
```diff
@@ -567,7 +572,21 @@ async def stream_async(
                     yield event["data"]
             ```
         """
-        callback_handler = kwargs.get("callback_handler", self.callback_handler)
+        merged_state = {}
```
Comment (afarntrog) [inline src/strands/agent/agent.py:576]: I think this code can be cleaned up a bit as it looks like we're setting `merged_state["invocation_state"] = invocation_state` regardless. Therefore, we should consider:
```python
merged_state = {}

if kwargs:
    logger.warning("`**kwargs` parameter is deprecated, use `invocation_state` instead.")
    merged_state.update(kwargs)

if invocation_state is not None:
    merged_state["invocation_state"] = invocation_state
```

Comment (JackYPCOnline) [reply src/strands/agent/agent.py:576]: I put it seprate because we are accepting both `invocation_sate` & `kwargs`, first, we want to check if user pass in `kwarg`, then check if they also pass in `invocation_state`. 
If no `kwargs` we update `invocation_state`, although result might be same, current approach is more readable and maintainable after we remove `kwargs`.

File: src/strands/agent/agent.py [hunk 9]
```diff
@@ -528,9 +534,7 @@ async def structured_output_async(self, output_model: Type[T], prompt: AgentInpu
                 self.hooks.invoke_callbacks(AfterInvocationEvent(agent=self))
 
     async def stream_async(
-        self,
-        prompt: AgentInput = None,
-        **kwargs: Any,
+        self, prompt: AgentInput = None, *, invocation_state: Optional[dict[str, Any]] = None, **kwargs: Any
```
Comment (Unshure) [inline src/strands/agent/agent.py]: nit
```suggestion
        self, prompt: AgentInput = None, *, invocation_state: dict[str, Any] | None = None, **kwargs: Any
```

File: tests/strands/agent/test_agent.py [hunk 10]
```diff
@@ -1877,3 +1877,17 @@ def test_tool(action: str) -> str:
     assert '"action": "test_value"' in tool_call_text
     assert '"agent"' not in tool_call_text
     assert '"extra_param"' not in tool_call_text
+
+
+def test_agent__call__handles_none_invocation_state(mock_model, agent):
```
Comment (Unshure) [inline tests/strands/agent/test_agent.py:1883]: Can you add one more unit test that can see what invocation state is passed to the event loop?

I want an invocation like this:
```
agent("hello!", invocation_state={"my": "state"}, other_kwarg="foobar")
```

To result in an invocation state like this:
```
event_loop_cycle.called_with(mock_agent, invocation_state=
{
  "invocation_state": {
    "my": "state"
  },
  "other_kwarg": "foobar"
})
```

File: src/strands/agent/agent.py [hunk 11]
```diff
@@ -567,7 +572,21 @@ async def stream_async(
                     yield event["data"]
             ```
         """
-        callback_handler = kwargs.get("callback_handler", self.callback_handler)
+        merged_state = {}
+        if kwargs:
+            logger.warning("`**kwargs` parameter is deprecated, use `invocation_state` instead.")
```
Comment (zastrowm) [inline src/strands/agent/agent.py]: Use python `warnings` for this:
https://github.com/zastrowm/sdk-python/blob/6ea8f723fdf2daafc8500fe6484510216495deef/src/strands/models/bedrock.py#L960 (and then when testing, use `captured_warnings` to verify it)

File: tests/strands/agent/test_agent.py (hunk 10 above)
Comment (zastrowm) [reply tests/strands/agent/test_agent.py:1883]: And similarly, 

```python
agent("hello!", invocation_state={"my": "state"})
```

called with:

```
event_loop_cycle.called_with(mock_agent, invocation_state={"my": "state"})
```

File: src/strands/agent/agent.py (hunk 6 above)
Comment (zastrowm) [reply src/strands/agent/agent.py]: I'm not sure I agree with this; see the Unit Test comment below.  But the intent is that users would replace `kwargs` with `invocation_state` - but if they're already using kwargs, then we mimic the old behavior to avoid a breaking change, while nudging them towards `invocation_state`

Basically it should be something like:

```python
if kwargs:
   merged_state = kwargs
   # emit warning
   if invocation_state:
       # If the user is using kwargs, that means they intended all arguments to be kwargs
       merged_state["invocation_state'] = kwargs
else:
    merged_state = invocation_state
```

File: tests/strands/agent/test_agent.py (hunk 10 above)
Comment (JackYPCOnline) [reply tests/strands/agent/test_agent.py:1883]: added.

File: src/strands/agent/agent.py (hunk 6 above)
Comment (JackYPCOnline) [reply src/strands/agent/agent.py]: Adjusted.
```
if xxx:
 ........
else:
    if invocation_state is not none: # Added this line, we are not accepting `None` now.
```

File: src/strands/agent/agent.py [hunk 12]
```diff
@@ -567,7 +573,21 @@ async def stream_async(
...
+                merged_state["invocation_state"] = invocation_state
+        else:
+            if invocation_state is not None:
+                merged_state = invocation_state
+
+        # Get callback handler from merged state or use default
+        invocation_state_dict = merged_state.get("invocation_state", {})
+        callback_handler = invocation_state_dict.get(
+            "callback_handler", merged_state.get("callback_handler", self.callback_handler)
+        )
```
Comment (Unshure) [inline src/strands/agent/agent.py]: I would say lets not support this feature in the new invocation_state change. This is a leftover from how we used to handle kwargs
```suggestion
        callback_handler = self.callback_handler
        if kwargs:
            callback_handler = kwargs.get("callback_handler", self.callback_handler)
```

File: tests/strands/agent/test_agent.py (hunk 10 above)
Comment (Unshure) [reply tests/strands/agent/test_agent.py:1883]: ^ Can you add the test case Mackenzie called out as well?

Comment (JackYPCOnline) [reply tests/strands/agent/test_agent.py:1883]: yep, see this [test](https://github.com/strands-agents/sdk-python/pull/966/files#diff-1cd3dfad7d3f3fbb13130bb7a6d06676c4f128a7bec43a7153691b1d24ba3af5R1927)

Comment (Unshure) [reply tests/strands/agent/test_agent.py:1883]: Ahh my bad!

File: src/strands/agent/agent.py [hunk 13]
```diff
@@ -528,9 +535,7 @@ async def structured_output_async(self, output_model: Type[T], prompt: AgentInpu
                 self.hooks.invoke_callbacks(AfterInvocationEvent(agent=self))
 
     async def stream_async(
-        self,
-        prompt: AgentInput = None,
-        **kwargs: Any,
+        self, prompt: AgentInput = None, *, invocation_state: dict[str, Any] | None = None, **kwargs: Any
```
Comment (dbschmigelski) [inline src/strands/agent/agent.py:538]: we should update our graph and potentially swarm and base to not use kwargs if we are doing this https://github.com/strands-agents/sdk-python/blob/main/src/strands/multiagent/graph.py#L575

otherwise graph users, using invocation state, will get this warning with no way to switch

Comment (JackYPCOnline) [reply src/strands/agent/agent.py:538]: I see just change `**invocation_state` to be `invocation_state = **invocation_state` fix this concern

Comment (dbschmigelski) [reply src/strands/agent/agent.py:538]: yep exactly

Comment (dbschmigelski) [reply src/strands/agent/agent.py:538]: well invocation_state=invocation_state no unpacking needed

File: src/strands/multiagent/base.py [hunk 14]
```diff
@@ -85,15 +86,14 @@ class MultiAgentBase(ABC):
 
     @abstractmethod
     async def invoke_async(
-        self, task: str | list[ContentBlock], invocation_state: dict[str, Any] | None = None, **kwargs: Any
+        self, task: str | list[ContentBlock], invocation_state: dict[str, Any] | None = None
```
Comment (dbschmigelski) [inline src/strands/multiagent/base.py]: sorry, we still want kwargs here to lets us expand the number of params in the future

File: src/strands/agent/agent.py [hunk 15]
```diff
@@ -567,7 +573,19 @@ async def stream_async(
...
+            merged_state.update(kwargs)
+            if invocation_state is not None:
+                merged_state["invocation_state"] = invocation_state
+        else:
+            if invocation_state is not None:
+                merged_state = invocation_state
+
+        callback_handler = self.callback_handler
+        if kwargs:
+            callback_handler = kwargs.get("callback_handler", self.callback_handler)
```
Comment (dbschmigelski) [inline src/strands/agent/agent.py:588]: nit: we could move this inside the kwargs if statement rather than doing another

File: src/strands/multiagent/base.py [hunk 16]
```diff
@@ -111,8 +112,12 @@ def __call__(
         if invocation_state is None:
             invocation_state = {}
 
+        if kwargs:
```
Comment (dbschmigelski) [inline src/strands/multiagent/base.py:115]: I'm not certain we need this warning since we still want kwargs to allow us to expand the number of params here. But I suppose its fine since graph and swarm will be the ones passing in the inputs.
